service/history/workflow/update/registry.go

```diff
@@ -26,229 +26,202 @@ package update
 
 import (
 	"context"
-	"fmt"
 	"sync"
 
-	"github.com/gogo/protobuf/types"
-	historypb "go.temporal.io/api/history/v1"
+	"go.opentelemetry.io/otel/trace"
 	protocolpb "go.temporal.io/api/protocol/v1"
-	"go.temporal.io/api/serviceerror"
 	updatepb "go.temporal.io/api/update/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
 )
 
 type (
 	Registry interface {
-		Add(request *updatepb.Request) (*Update, Duplicate, RemoveFunc)
+		// FindOrCrate finds an existing Update or creates a new one. The second
+		// return value (bool) indicates whether the Update returned is newly
+		// created (true) for an existing update (false).
+		FindOrCreate(ctx context.Context, protocolInstanceID string) (*Update, bool, error)
 
+		// Find finds an existing update in this Registry but does not create a
+		// new update it is absent.
+		Find(ctx context.Context, protocolInstanceID string) (*Update, bool)
+
+		// CreateOutgoingMessages polls each registered Update for outbound
+		// messages and returns them.
 		CreateOutgoingMessages(startedEventID int64) ([]*protocolpb.Message, error)
```

Comment:
startedEventID probably shouldn't be here - it breaks layering - but it's fine for now.